src/Cron.php

```diff
@@ -131,4 +156,38 @@ public static function isValid(string $expression): bool
     {
         return CronExpression::isValidExpression($expression);
     }
+
+    public static function fromJson(string $data): Frequency
+    {
+        $data = json_decode($data, true);
+        if (! isset($data['expression'])) {
+            throw new InvalidArgumentException('Missing required attribute "expression"');
+        }
+
+        $self = new static($data['expression']);
+        if (isset($data['start'])) {
+            $self->startAt(new DateTime($data['start']));
+        }
+
+        if (isset($data['end'])) {
+            $self->endAt(new DateTime($data['end']));
+        }
+
+        return $self;
+    }
+
+    #[\ReturnTypeWillChange]
+    public function jsonSerialize()
+    {
+        $parts = ['expression' => $this->getExpression()];
+        if ($this->start) {
+            $parts['start'] = $this->start->format(DateTimeInterface::ATOM);
```

Comment:
I would use a custom format, i.e. `'Y-m-d\TH:i:s.uP'`:

https://3v4l.org/n2EVk#veol

@nilmerg Do you agree?

Reply:
Why?

Reply:
> I would use a custom format, i.e. `'Y-m-d\TH:i:s.uP'`:

None of the frequencies actually use this additional information.
https://github.com/Icinga/ipl-scheduler/blob/95b9171ca77654883c53681d31bc00ce08bc960d/src/RRule.php#L165-L168

Reply:
To put it simply, the serialized data is wrong. This is a library that also serves as a best practice bullet-proof implementation that should produce the same datetime as it was before.

Reply:
Ah, because milliseconds aren't preserved. If you insist :man_shrugging: . I only would care if the scheduler ticks even fast enough to respect those. Does it?

Reply:
It [does](https://github.com/reactphp/event-loop#addtimer). However, my main argument is that if we need to serialize datetimes to JSON anywhere else again, we might look at the implementation here and use that as a reference. So it has to be 100% correct.